src/v/raft/append_entries_buffer.cc

```diff
@@ -0,0 +1,121 @@
+#include <variant>
+#include <vector>
+namespace raft {
+
+append_entries_buffer::append_entries_buffer(
+  consensus& c, size_t max_buffered_elements)
+  : _consensus(c)
+  , _max_buffered(max_buffered_elements) {}
+
+ss::future<append_entries_reply>
+append_entries_buffer::enqueue(append_entries_request&& r) {
+    return ss::with_gate(_gate, [this, r = std::move(r)]() mutable {
+        // we normally do not want to wait as it would cause requests
+        // reordering. Reordering may only happend if we would wait on condition
+        // variable.
+
+        return _flushed
+          .wait([this] { return _requests.size() < _max_buffered; })
+          .then([this, r = std::move(r)]() mutable {
+              ss::promise<append_entries_reply> p;
+              auto f = p.get_future();
+              _requests.push_back(std::move(r));
+              _responses.push_back(std::move(p));
+              _enqueued.signal();
+              return f;
+          });
+    });
+}
+
+ss::future<> append_entries_buffer::stop() {
+    _enqueued.broken();
+    _flushed.broken();
+    return _gate.close();
+}
+
+void append_entries_buffer::start() {
+    (void)ss::with_gate(_gate, [this]() mutable {
+        return ss::do_until(
+          [this] { return _gate.is_closed(); },
+          [this]() mutable {
+              return _enqueued.wait([this] { return !_requests.empty(); })
+                .then([this]() mutable { return flush(); })
+                .handle_exception_type(
+                  [](const ss::broken_condition_variable&) {
+                      // ignore exception, we are about to stop
+                  });
+          });
+    });
+}
+
+ss::future<> append_entries_buffer::flush() {
+    auto requests = std::exchange(_requests, {});
+    auto response_promises = std::exchange(_responses, {});
+
+    return _consensus._op_lock.with(
+      [this,
+       requests = std::move(requests),
+       response_promises = std::move(response_promises)]() mutable {
+          return do_flush(std::move(requests), std::move(response_promises));
+      });
+}
+
+ss::future<> append_entries_buffer::do_flush(
+  request_t requests, response_t response_promises) {
+    bool needs_flush = false;
+    std::vector<reply_t> replies;
+    replies.reserve(requests.size());
+    for (auto& req : requests) {
+        if (req.flush) {
+            needs_flush = true;
+        }
+        try {
+            // NOTE: do_append_entries do not flush
+            auto reply = co_await _consensus.do_append_entries(std::move(req));
+            replies.emplace_back(reply);
+        } catch (...) {
+            replies.emplace_back(std::current_exception());
+        }
+    }
+    if (needs_flush) {
+        co_await _consensus.flush_log();
+    }
+
+    propagate_results(replies, std::move(response_promises));
+    _flushed.broadcast();
+    co_return;
+}
+
+void append_entries_buffer::propagate_results(
+  std::vector<reply_t>& replies, response_t response_promises) {
```

Comment:
is there a reason why both parameters are not taken by value?

Reply:
good point, this is leftover from future based implementation